File: experiments/engine-kt/provider-compatible/src/main/kotlin/com/garive/eng/kt/provider/compatible/Deployment.kt
package com.garive.eng.kt.provider.compatible

import com.garive.eng.kt.anthropic.DocumentSource
import com.garive.eng.kt.anthropic.ImageSource
import com.garive.eng.kt.anthropic.ThinkingConfig
import com.garive.eng.kt.llm.InterruptionKind
import com.garive.eng.kt.llm.ModelCapability
import com.garive.eng.kt.llm.RejectionKind
import com.garive.eng.kt.llm.UnavailableKind
import com.garive.eng.kt.openai.ImageDetail
import com.garive.eng.kt.openai.ReasoningConfig

/** Immutable configuration for one Responses-compatible deployment. */
public data class ResponsesDeployment(
    public val targetId: String,
    public val modelId: String,
    public val capabilities: Set<ModelCapability>,
    public val defaultMaxOutputTokens: ULong? = null,
    public val mediaBindings: Map<String, ResponsesMediaBinding> = emptyMap(),
    public val reasoning: ReasoningConfig? = null,
    public val errorPolicy: ProtocolErrorPolicy = ProtocolErrorPolicy.empty(),
)

/** Explicit binding for one neutral Responses media reference. */
public sealed interface ResponsesMediaBinding {
    /** URL or data-URL image binding. */
    public data class Url(public val value: String, public val detail: ImageDetail? = null) : ResponsesMediaBinding
    /** Previously uploaded protocol file binding. */
    public data class FileId(public val value: String, public val detail: ImageDetail? = null) : ResponsesMediaBinding
}

/** Immutable configuration for one Messages-compatible deployment. */
public data class MessagesDeployment(
    public val targetId: String,
    public val modelId: String,
    public val capabilities: Set<ModelCapability>,
    public val defaultMaxOutputTokens: ULong? = null,
    public val mediaBindings: Map<String, MessagesMediaBinding> = emptyMap(),
    public val thinking: ThinkingConfig? = null,
    public val errorPolicy: ProtocolErrorPolicy = ProtocolErrorPolicy.empty(),
)

/** Explicit binding for one neutral Messages media reference. */
public sealed interface MessagesMediaBinding {
    /** Official image-source binding. */
    public data class Image(public val source: ImageSource) : MessagesMediaBinding
    /** Official document-source binding. */
    public data class Document(public val source: DocumentSource) : MessagesMediaBinding
}

/** Exact protocol error identity; human-readable messages are deliberately absent. */
public data class ErrorSignature(
    public val status: UShort,
    public val protocolType: String,
    public val code: String?,
)

/** Provider-neutral disposition for one exact protocol error signature. */
public sealed interface ErrorDisposition {
    /** Rejected invocation. */
    public data class Rejected(public val kind: RejectionKind) : ErrorDisposition
    /** Temporarily unavailable invocation. */
    public data class Unavailable(public val kind: UnavailableKind) : ErrorDisposition
    /** Interrupted invocation. */
    public data class Interrupted(public val kind: InterruptionKind) : ErrorDisposition
}

/** Immutable exact-match protocol error policy. */
public class ProtocolErrorPolicy private constructor(
    private val mappings: Map<ErrorSignature, ErrorDisposition>,
) {
    /** Returns the disposition for an exact signature. */
    public fun classify(signature: ErrorSignature): ErrorDisposition? = mappings[signature]

    public companion object {
        /** Creates an empty policy. */
        public fun empty(): ProtocolErrorPolicy = ProtocolErrorPolicy(emptyMap())

        /** Creates a policy and rejects duplicate signatures. */
        public fun of(rules: List<Pair<ErrorSignature, ErrorDisposition>>): ProtocolErrorPolicy {
            require(rules.map(Pair<ErrorSignature, ErrorDisposition>::first).distinct().size == rules.size) {
                "duplicate error signature"
            }
            return ProtocolErrorPolicy(rules.toMap())
        }
    }
}

File: providers/compatible/src/deployment.rs
use std::collections::{BTreeMap, BTreeSet};

use garive_anthropic_messages::{DocumentSource, ImageSource, ThinkingConfig};
use garive_llm::{InterruptionKind, ModelCapability, RejectionKind, UnavailableKind};
use garive_openai_responses::{ImageDetail, ReasoningConfig};

/// Immutable configuration for one Responses-compatible deployment.
#[derive(Clone, Debug, PartialEq)]
pub struct ResponsesDeployment {
    /// Exact neutral target admitted by this deployment.
    pub target_id: String,
    /// Protocol model identifier selected by configuration.
    pub model_id: String,
    /// Capabilities explicitly admitted by the deployment.
    pub capabilities: BTreeSet<ModelCapability>,
    /// Optional configured output-token default.
    pub default_max_output_tokens: Option<u64>,
    /// Neutral media references resolved to protocol image inputs.
    pub media_bindings: BTreeMap<String, ResponsesMediaBinding>,
    /// Optional protocol reasoning configuration.
    pub reasoning: Option<ReasoningConfig>,
    /// Exact error classification policy.
    pub error_policy: ProtocolErrorPolicy,
}

/// Explicit binding for one neutral Responses media reference.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ResponsesMediaBinding {
    /// Bind to a URL or data URL.
    Url {
        /// URL value.
        value: String,
        /// Optional image fidelity.
        detail: Option<ImageDetail>,
    },
    /// Bind to an already uploaded protocol file.
    FileId {
        /// Protocol file identity.
        value: String,
        /// Optional image fidelity.
        detail: Option<ImageDetail>,
    },
}

/// Immutable configuration for one Messages-compatible deployment.
#[derive(Clone, Debug, PartialEq)]
pub struct MessagesDeployment {
    /// Exact neutral target admitted by this deployment.
    pub target_id: String,
    /// Protocol model identifier selected by configuration.
    pub model_id: String,
    /// Capabilities explicitly admitted by the deployment.
    pub capabilities: BTreeSet<ModelCapability>,
    /// Optional configured output-token default.
    pub default_max_output_tokens: Option<u64>,
    /// Neutral media references resolved to protocol content.
    pub media_bindings: BTreeMap<String, MessagesMediaBinding>,
    /// Optional protocol thinking configuration.
    pub thinking: Option<ThinkingConfig>,
    /// Exact error classification policy.
    pub error_policy: ProtocolErrorPolicy,
}

/// Explicit binding for one neutral Messages media reference.
#[derive(Clone, Debug, PartialEq)]
pub enum MessagesMediaBinding {
    /// Bind to an official image source.
    Image(ImageSource),
    /// Bind to an official document source.
    Document(DocumentSource),
}

/// Exact protocol error identity; message text is deliberately excluded.
#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct ErrorSignature {
    /// HTTP status returned by Runtime transport.
    pub status: u16,
    /// Exact protocol error type.
    pub protocol_type: String,
    /// Exact optional protocol error code.
    pub code: Option<String>,
}

/// Provider-neutral disposition assigned to an exact error signature.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ErrorDisposition {
    /// Reject as authentication or authorization failure.
    Rejected(RejectionKind),
    /// Report temporary unavailability.
    Unavailable(UnavailableKind),
    /// Report an interrupted invocation.
    Interrupted(InterruptionKind),
}

/// Immutable table of admitted protocol error mappings.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct ProtocolErrorPolicy {
    mappings: BTreeMap<ErrorSignature, ErrorDisposition>,
}

impl ProtocolErrorPolicy {
    /// Constructs a policy from exact signatures; duplicate signatures are overwritten last.
    pub fn new(mappings: impl IntoIterator<Item = (ErrorSignature, ErrorDisposition)>) -> Self {
        Self {
            mappings: mappings.into_iter().collect(),
        }
    }

    /// Looks up an exact signature without inspecting human-readable messages.
    pub fn classify(&self, signature: &ErrorSignature) -> Option<ErrorDisposition> {
        self.mappings.get(signature).copied()
    }
}
